### PR TITLE
Performance improvement/runset load time and restore activity

### PR DESCRIPTION
**Pull request by vyasprateek**

### CODE REVIEW

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs [hunk 1]
```diff
@@ -1091,18 +1096,73 @@ private void xSaveBusinessFlowBtn_Click(object sender, RoutedEventArgs e)
...
             {
-                mActivitiesPage.ListView.UpdateGrouping();
-                mBusinessFlow.SaveBackup();
+                xItemsTabsSection.Visibility = Visibility.Collapsed;
+                xItemsLoadingPnl.Visibility = Visibility.Visible;
+                SwapLoadingPrefixText("Undoing", false);
+
+                try
+                {
+                    //Mouse.OverrideCursor = System.Windows.Input.Cursors.Wait;
```
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1146]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174&bid=19456803)

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs [hunk 2]
```diff
@@ -137,6 +138,15 @@ public void AddRepositoryItem(RepositoryItemBase repositoryItem)
             SRII.ItemRootRepositoryFolder.AddRepositoryItem(repositoryItem);
         }
 
+        public async Task SaveRepositoryItemAsync(RepositoryItemBase repositoryItem)
+        {
+            repositoryItem.OngoingAsyncSaveProcess = true;
+
+            await Task.Run(() => SaveRepositoryItem(repositoryItem));
```
Comment (ambarkhan) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:145]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs [hunk 3]
```diff
@@ -1091,18 +1096,73 @@ private void xSaveBusinessFlowBtn_Click(object sender, RoutedEventArgs e)
...
+                }
+                catch (Exception ex)
+                {
+                    Reporter.ToUser(eUserMsgKey.StaticWarnMessage, string.Format("Failed to undo changes to the item '{0}', please view log for more details", mBusinessFlow.ItemName));
+                    Reporter.ToLog(eLogLevel.ERROR, string.Format("Failed to undo changes to the item '{0}'", mBusinessFlow.ItemName), ex);
+                }
+                finally
+                {
+                    Reporter.HideStatusMessage();
+                    //Mouse.OverrideCursor = null;
```
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1161]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs [hunk 4]
```diff
@@ -1076,6 +1073,18 @@ public void SetDirtyStatusToNoChange()
             DirtyStatus = eDirtyStatus.NoChange;
 
             // Properties
+            //foreach (PropertyInfo PI in this.GetType().GetProperties().Where(pi =>
+            //    pi.GetCustomAttribute(typeof(IsSerializedForLocalRepositoryAttribute)) != null
+            //        && typeof(IObservableList).IsAssignableFrom(pi.PropertyType)
+            //            && (pi.GetValue(this) as IObservableList) != null))
+            //{
```
Comment (ambarkhan) [inline Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs:1080]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs [hunk 5]
```diff
@@ -1091,18 +1096,73 @@ private void xSaveBusinessFlowBtn_Click(object sender, RoutedEventArgs e)
...
-                mBusinessFlow.SaveBackup();
+                xItemsTabsSection.Visibility = Visibility.Collapsed;
+                xItemsLoadingPnl.Visibility = Visibility.Visible;
+                SwapLoadingPrefixText("Undoing", false);
+
+                try
+                {
+                    //Mouse.OverrideCursor = System.Windows.Input.Cursors.Wait;
+                    Reporter.ToStatus(eStatusMsgKey.StaticStatusProcess, null, string.Format("Undoing changes for '{0}'...", mBusinessFlow.ItemName));
+                    await Task.Run(() => mBusinessFlow.RestoreFromBackup(true, true));
```
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1148]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174&bid=19456803)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs (hunk 1 above)
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1146]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs (hunk 2 above)
Comment (ambarkhan) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:145]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs (hunk 4 above)
Comment (ambarkhan) [inline Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs:1080]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs (hunk 5 above)
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1148]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs (hunk 3 above)
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1161]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs [hunk 6]
```diff
@@ -1091,18 +1096,73 @@ private void xSaveBusinessFlowBtn_Click(object sender, RoutedEventArgs e)
 
             Reporter.ToStatus(eStatusMsgKey.SaveItem, null, mBusinessFlow.Name,
                                       GingerDicser.GetTermResValue(eTermResKey.BusinessFlow));
-            WorkSpace.Instance.SolutionRepository.SaveRepositoryItem(mBusinessFlow);
+
+            SwapLoadingPrefixText("Saving", false);
+            xItemsTabsSection.Visibility = Visibility.Collapsed;
+            xItemsLoadingPnl.Visibility = Visibility.Visible;
+
```
Comment (MeniKadosh1) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1103]: do try-finally

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs [hunk 7]
```diff
@@ -1091,18 +1096,73 @@ private void xSaveBusinessFlowBtn_Click(object sender, RoutedEventArgs e)
...
+            if (Reporter.ToUser(eUserMsgKey.AskIfToUndoItemChanges, mBusinessFlow.ItemName) == eUserMsgSelection.Yes)
             {
-                mActivitiesPage.ListView.UpdateGrouping();
-                mBusinessFlow.SaveBackup();
+                xItemsTabsSection.Visibility = Visibility.Collapsed;
+                xItemsLoadingPnl.Visibility = Visibility.Visible;
+                SwapLoadingPrefixText("Undoing", false);
+
+                try
+                {
```
Comment (MeniKadosh1) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1145]: post merge with Meni branch- need to copy here the validation added to check if backup exist and not in progress

File: Ginger/Ginger/Ginger.csproj [hunk 8]
```diff
@@ -342,6 +342,7 @@
       <Private>True</Private>
       <Private>True</Private>
     </Reference>
+    <Reference Include="System.ServiceModel" />
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Ginger.csproj:345]: need to check what are those changes for and if we need to merge them to master

File: Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs [hunk 9]
```diff
@@ -73,6 +73,8 @@ public abstract class RepositoryItemBase : INotifyPropertyChanged, ISearchFilter
         //DO Not save
         protected ConcurrentDictionary<string, object> mBackupDic;
 
+        public bool OngoingAsyncSaveProcess = false;
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs:76]: need to remove if not needed

File: Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs [hunk 10]
```diff
@@ -1076,6 +1073,18 @@ public void SetDirtyStatusToNoChange()
             DirtyStatus = eDirtyStatus.NoChange;
 
             // Properties
+            //foreach (PropertyInfo PI in this.GetType().GetProperties().Where(pi =>
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs:1076]: remove comments

File: Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs [hunk 11]
```diff
@@ -1092,6 +1101,17 @@ public void SetDirtyStatusToNoChange()
             }
 
             // Fields
+            //foreach (FieldInfo FI in this.GetType().GetFields().Where(fi =>
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs:1104]: remove comments

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs (hunk 2 above)
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:145]: need to test with .ConfigureAwait(false)

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs [hunk 12]
```diff
@@ -150,7 +160,14 @@ public void SaveRepositoryItem(RepositoryItemBase repositoryItem)
 
             repositoryItem.UpdateBeforeSave();
 
+            DateTime startTime = DateTime.Now;
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:163]: clean all time checking code

File: Ginger/GingerCoreNET/Run/RunsetExecutor.cs [hunk 13]
```diff
@@ -168,7 +168,11 @@ public void InitRunner(GingerRunner runner)
             ObservableList<BusinessFlow> runnerFlows = new ObservableList<BusinessFlow>();
             foreach (BusinessFlowRun businessFlowRun in runner.BusinessFlowsRunList.ToList())
             {
+                DateTime strtTime = DateTime.Now;
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/Run/RunsetExecutor.cs:171]: clean time flags

File: Ginger/GingerCoreNETUnitTest/SolutionRepositoryLib/RepositoryTest.cs [hunk 14]
```diff
@@ -1134,7 +1134,89 @@ public void ActivityVariableDependancyTest_WithCreateInstance()
...
+            activity.Acts.Add(act2);
+
+            bf.Activities.RemoveAt(0);
+            bf.Activities.Add(activity);
+            bf.Activities.Add(activity2);
 
+            bf.RepositorySerializer.SaveToFile(bf, TestResources.GetTempFile("BF.xml"));
+
+            //Act
+            BusinessFlow copiedItem = (BusinessFlow)bf.CreateCopy();
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNETUnitTest/SolutionRepositoryLib/RepositoryTest.cs:1166]: need extra test to check that also changes done after save been copied

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174&bid=19456803)

File: Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs (hunk 4 above)
Comment (ambarkhan) [inline Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs:1080]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs (hunk 2 above)
Comment (ambarkhan) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:145]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs [hunk 15]
```diff
@@ -1089,20 +1106,73 @@ private void xSaveBusinessFlowBtn_Click(object sender, RoutedEventArgs e)
...
+            if (Reporter.ToUser(eUserMsgKey.AskIfToUndoItemChanges, mBusinessFlow.ItemName) == eUserMsgSelection.Yes)
             {
-                mActivitiesPage.ListView.UpdateGrouping();
-                mBusinessFlow.SaveBackup();
+                SwapLoadingPrefixText("Undoing", false);
+
+                try
+                {
+                    Reporter.ToStatus(eStatusMsgKey.StaticStatusProcess, null, string.Format("Undoing changes for '{0}'...", mBusinessFlow.ItemName));
+                    await Task.Run(() => mBusinessFlow.RestoreFromBackup(true, true));
```
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1161]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs (hunk 6 above)
Comment (vyasprateek) [reply Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs]: Done

File: Ginger/Ginger/Ginger.csproj (hunk 8 above)
Comment (vyasprateek) [reply Ginger/Ginger/Ginger.csproj]: Removed the unnecessary/accidentally added refs.

File: Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs (hunk 9 above)
Comment (vyasprateek) [reply Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs:76]: Removed. No more required.

File: Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs [hunk 16]
```diff
@@ -287,7 +287,7 @@ public void CreateBackup(bool isLocalBackup = false)
             });
         }
 
-        private bool IsDoNotBackupAttr(MemberInfo mi)
+        public bool IsDoNotBackupAttr(MemberInfo mi)
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/Repository/RepositoryItemBase.cs:290]: return to private

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs [hunk 17]
```diff
@@ -160,9 +173,21 @@ public void SaveRepositoryItem(RepositoryItemBase repositoryItem)
             RefreshParentFoldersSoucerControlStatus(Path.GetDirectoryName(repositoryItem.FilePath));
 
             if (repositoryItem.DirtyStatus != Common.Enums.eDirtyStatus.NoTracked)
+            {
+                startTime = DateTime.Now;
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:177]: clean time tracking

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs [hunk 18]
```diff
@@ -160,9 +173,21 @@ public void SaveRepositoryItem(RepositoryItemBase repositoryItem)
...
+
+                timeTaken = (DateTime.Now - startTime).TotalSeconds;
+                Reporter.ToLog(eLogLevel.INFO, "repositoryItem.SetDirtyStatusToNoChange() took " + timeTaken + " seconds");
+            }
+
+            startTime = DateTime.Now;
+
             repositoryItem.CreateBackup();
-        }      
+            timeTaken = (DateTime.Now - startTime).TotalSeconds;
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:188]: clean time tracking

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs (hunk 15 above)
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1161]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs [hunk 19]
```diff
@@ -137,6 +138,11 @@ public void AddRepositoryItem(RepositoryItemBase repositoryItem)
             SRII.ItemRootRepositoryFolder.AddRepositoryItem(repositoryItem);
         }
 
+        public async Task SaveRepositoryItemAsync(RepositoryItemBase repositoryItem)
+        {
+            await Task.Run(() => SaveRepositoryItem(repositoryItem));
```
Comment (ambarkhan) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:143]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174&bid=19456803)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174&bid=19456803)

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs (hunk 19 above)
Comment (ambarkhan) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:143]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs [hunk 20]
```diff
@@ -1089,20 +1106,73 @@ private void xSaveBusinessFlowBtn_Click(object sender, RoutedEventArgs e)
...
```
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1161]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174&bid=19456803)

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs (hunk 20 above)
Comment (ambarkhan) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:1161]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

File: Ginger/GingerCoreCommon/Repository/SolutionRepository.cs (hunk 19 above)
Comment (ambarkhan) [inline Ginger/GingerCoreCommon/Repository/SolutionRepository.cs:143]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=5936174)

**Pull request closed without merging**